src/nbl/video/utilities/CComputeBlit.cpp

```diff
@@ -0,0 +1,152 @@
+		"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
+		"#define _NBL_GLSL_BLIT_DIM_COUNT_ %d\n"
+		"#define _NBL_GLSL_BLIT_ALPHA_BIN_COUNT_ %d\n"
+		"#define _NBL_GLSL_BLIT_NORMALIZATION_OUT_IMAGE_FORMAT_ %s\n"
+		"%s\n" // _NBL_GLSL_BLIT_NORMALIZATION_SOFTWARE_CODEC_
+		"%s\n", // Todo(achal): Remove format include in favour of the symbol gotten from outImageFormatGLSLString
+		inImageType >= asset::IImage::ET_1D ? workgroupDims.x : 1u,
+		inImageType >= asset::IImage::ET_2D ? workgroupDims.y : 1u,
+		inImageType >= asset::IImage::ET_3D ? workgroupDims.z : 1u,
+		static_cast<uint32_t>(inImageType) + 1u,
+		alphaBinCount,
+		glslFormatQualifier,
+		outImageFormat != outImageViewFormat ? "#define _NBL_GLSL_BLIT_NORMALIZATION_SOFTWARE_CODEC_\n" : "",
+		reinterpret_cast<const char*>(formatInclude));
+
+	auto gpuUnspecShader = device->createGPUShader(std::move(cpuShaderOverriden));
+
+	return device->createGPUSpecializedShader(gpuUnspecShader.get(), { nullptr, nullptr, "main" });
+}
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createBlitSpecializedShader(const asset::E_FORMAT inImageFormat, const asset::E_FORMAT outImageFormat,
+	const asset::E_FORMAT outImageViewFormat, const asset::IImage::E_TYPE inImageType, const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent,
+	const asset::IBlitUtilities::E_ALPHA_SEMANTIC alphaSemantic, const uint32_t workgroupSize, const uint32_t alphaBinCount)
+{
+	char shaderStart[] = "#version 460 core\n#include \"nbl/builtin/glsl/blit/default_compute_blit.comp\"\n";
+	auto shader = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(strlen(shaderStart), shaderStart, core::adopt_memory);
+
+	auto cpuShader = core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shader), asset::ICPUShader::buffer_contains_glsl, asset::IShader::ESS_COMPUTE, "CComputeBlit::createBlitSpecializedShader");
+	
+	const core::vectorSIMDu32 windowDim = getWindowDim(inExtent, outExtent);
+
+	// Fail if any dimension of the window is bigger than workgroup size, forcing us to reuse inovcation to process just a single dimension of that window
+	if ((windowDim.x > workgroupSize) || (windowDim.y > workgroupSize) || (windowDim.z > workgroupSize))
+		return nullptr;
+
+	const uint32_t windowPixelCount = windowDim.x * windowDim.y * windowDim.z;
+	// Fail if I would need to reuse invocations just to process a single window
+	if (windowPixelCount > workgroupSize)
+		return nullptr;
+
+	const uint32_t outChannelCount = asset::getFormatChannelCount(outImageFormat);
+	// It is important not to use asset::getTexelOrBlockBytesize here.
+	// Even though input pixels are stored in the shared memory we use outChannelCount here because only outChannelCount channels of
+	// the input image actually need blitting and hence only those channels are stored in shared memory.
+	const size_t windowSize = static_cast<size_t>(windowPixelCount) * outChannelCount * sizeof(float);
```

Comment:
P.S. it will ever be below 32 floats if you enable the alpha coverage thing, because `workgroupInclusiveAdd` requires O(2*workgroup_size) in scratch size

Reply:
P.P.S. eh, it will always be above 32 floats with a sane workgroup size, cause window is at least 1x1

So just round up to next multiple of 64 floats.

Reply:
You can send the window DIMENSIONS (X,Y,Z) as push constants, but TOTAL WINDOW MEMORY SIZE needs to remain a `#define` but can be an upper bound (i.e. max against the `workgroupInclusiveAdd` scratch requirement [2x], then round up to next 64 dwords depending on size after maxxing)

So you can actually just do the loops on push constants
```glsl
for (uint i = 0; i < pc.windowDim.x; ++i)
{
    // do the thing
}
```
sacrificing unrolling for less pipeline combos and ability to cache them

Reply:
Resolved.